### ISSUE TITLE
[tests/vm] Add `dart_api_dl.h` expect tests

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#51467]
We should add some kind of tests to check ABI compatibility between different commits on the Dart SDK on `dart_api_dl.h`.

Some ideas on how to generate an expect file:

1. Use FFIgen's frontend to get the information from the AST.
   a. pros: configurable, no new dependencies
   b. cons: requires building some manual tooling
2. Use `buildtools/linux-x64/clang/bin/clang -x c-header -extract-api runtime/include/dart_api_dl.h -o output.json`
   a. non-solution: this information does not expand the types, just a shallow comparison, wouldn't have caught https://github.com/dart-lang/sdk/issues/51459
3. Use `buildtools/linux-x64/clang/bin/clang --preprocess runtime/include/dart_api_dl.h &> output.txt`
    a. cons: how to parse the preprocessor output?
    b. pros: no new dependencies
4. Use https://github.com/lvc/abi-compliance-checker
    a. pros: seems to be geared close to our use case, detects ABI incompatibilities between two versions of
    b. cons: works on binaries, we need to put a compiled binary in our DEPS and roll it on updates

steps for 4 :

```
dart-sdk/sdk$ buildtools/linux-x64/clang/bin/clang -shared -fPIC  runtime/include/dart_api_dl.c  -o runtime/lib/libdart_api_dl.so

master-dart-sdk/sdk$ buildtools/linux-x64/clang/bin/clang -shared -fPIC  runtime/include/dart_api_dl.c  -o runtime/lib/libdart_api_dl.so

dart-sdk/sdk$ cat NEW.xml 
<version>
    1.1
</version>
<headers>
    /usr/local/google/home/dacoharkes/master-dart-sdk/sdk/runtime/include/dart_api_dl.h
</headers>
<libs>
    /usr/local/google/home/dacoharkes/master-dart-sdk/sdk/runtime/lib/
</libs>

dart-sdk/sdk$ cat OLD.xml 
<version>
    1.0
</version>
<headers>
    /usr/local/google/home/dacoharkes/dart-sdk/sdk/runtime/include/dart_api_dl.h
</headers>
<libs>
    /usr/local/google/home/dacoharkes/dart-sdk/sdk/runtime/lib/
</libs>

dart-sdk/sdk$ abi-compliance-checker -lib NAME -old OLD.xml -new NEW.xml
Preparing, please wait ...
Using GCC 12 (x86_64-linux-gnu, target: x86_64)
WARNING: May not work properly with GCC 4.8.[0-2], 6.* and higher due to bug #78040 in GCC. Please try other GCC versions with the help of --gcc-path=PATH option or create ABI dumps by ABI Dumper tool instead to avoid using GCC. Test selected GCC version first by -test and -gcc-path options.
Checking header(s) 1.0 ...
Checking header(s) 1.1 ...
Comparing ABIs ...
Comparing APIs ...
Creating compatibility report ...
Binary compatibility: 97.1%
Source compatibility: 100%
Total binary compatibility problems: 5, warnings: 0
Total source compatibility problems: 0, warnings: 0
Report: compat_reports/NAME/1.0_to_1.1/compat_report.html
```

HTML report:
<img width="934" alt="image" src="https://user-images.githubusercontent.com/3601775/220099225-abe07634-d795-4914-b7ae-4158514af6bb.png">

@mkustermann wdyt about the abi-compliance-checker to address this?

Adding to an enum with a last entry item such as `Dart_CObject_kNumberOfTypes` will always flag an incompatibility when adding  one-before-last entry. But maybe that's useful for us to force us to review the change.

This would catch:

* https://github.com/dart-lang/sdk/issues/51459
* https://github.com/dart-lang/sdk/issues/47649

**Comment by mkustermann** [comment]
If we can avoid extra 3rd party dependencies, that would be good.

/cc @mraleph recently wrote some python scripts using clang to extract information from VM sources. May also be useful here.

**Comment by dcharkes** [comment]
cc @rmacnak-google

> I'm okay with this change, but assert there was no breaking change as the VM embedding API does not provide any stability guarantees.

We don't for the VM embedding AP, but we do for `dart_api_dl.h` which shares some types/functions.

**Comment by dcharkes** [comment]
I recently also found https://sourceware.org/libabigail/ as a more modern variant for a checker.

Step 1 is to get a working prototype with one of tools or a custom tool.
Step 2 is alignment on adding a new third-party dependency. If we use it for the embedder API checks, we should use the same one for Flutter's embedder API. cc @loic-sharma @jtmcdole

**Comment by ankitranjandev** [comment]
@dcharkes I would like to work on this

**Comment by dcharkes** [comment]
> @dcharkes I would like to work on this

@ankitranjandev This task requires (1) do research on what tool we should use and (2) then getting alignment with the stakeholders to get agreement on adding such tool.

First there is a question do we want only API-stability, or also ABI-stability.

For embedders of Dart, such as the Flutter engine, it is likely that dart_api.h is rolled in together with a new binary of the Dart VM. So for that we only need API stability. (pointed out by @mraleph recently)

For dynamic libraries produced in build hooks that are built against dart_api_dl.h we currently don't communicate API version. And dylibs might be precompiled and hosted online. So for these we'd want ABI compatibility. (And even if we were to start adding dart_api_dl.h and a version to the CodeConfig in the BuildInput, we would need to check that on any PR that does _not_ bump `DART_API_DL_MAJOR_VERSION` we have ABI compatibility.)

So the next steps here are: Explore what tools for ABI and API compatibility are available. Explore how to integrate those in the Dart CI. Write up pros and cons of different tools and different integration approaches on the Dart CI. And report that back here on this issue to gather alignment of all stake holders. (So, this is not really a coding, getting CI green, and landing the change task. More of an exploration and coming up with a proposal task.) Edit: We're happy receiving such kinds of contributions as well, I just wanted to make sure you know what you're getting in to! 😄

**Comment by ankitranjandev** [comment]
Thanks @dcharkes  for taking your time to provide me clarity on this issue.

I created a PoC using  https://sourceware.org/libabigail/ 
https://github.com/ankitranjandev/dart_abi_poc

I added checks on Github Actions which ensures the abi stability on a new MR comparing it with head branch.

It compiles a mock native extension using dart_api_dl.h with DWARF debug info and flags breaking ABI changes (e.g., modified signatures) using abidiff.

here's the pull request that breaks due to indirect sub-type change
https://github.com/ankitranjandev/dart_abi_poc/pull/1

Can you please check if it is of any help.

**Comment by dcharkes** [comment]
Nice!

We don't develop on GitHub, so we'll have to integrate this into a test that runs on Gerrit https://dart-review.googlesource.com/

I do really like the idea of downloading from an existing place. We could check against the last checked in version of the SDK:

https://github.com/dart-lang/sdk/blob/74a08cddac876d709225e45c2480967f3f898493/DEPS#L50-L58

The checked in SDK can be accessed via

https://github.com/dart-lang/sdk/blob/74a08cddac876d709225e45c2480967f3f898493/runtime/tests/vm/dart/use_flag_test_helper.dart#L64-L70

Could you try to see if you could write a Dart test using `libabigail` with that. Something like `runtime/tests/vm/dart/abi/dart_api_test.dart` and `runtime/tests/vm/dart/abi/dart_api_dl_test.dart`.

I think we should pull in `libabigail` via CIPD via the DEPS file. Let me see if I can get the right stakeholders on board to do that.

**Comment by dcharkes** [comment]
TIL that the new tool on the block is STG: https://android.googlesource.com/platform/external/stg/

@ankitranjandev I think it's probably an easier sell to add STG as a dependency than libabigail, because STG is from Google, which prevents us having to vet it as a third-party dependency.

**Comment by ankitranjandev** [comment]
@dcharkes 
Yes, STG is a successor to libabigail and is being actively maintained. Thank you for helping me with that info. I'll take a look into it.

So the action items are:
1. Create a PoC patchset on Gerrit with the last checked-in version of the SDK.
2. Use STG instead of libabigail, as it is more modern and maintained by Google, meaning fewer third-party dependencies.
3. Write test(s) to demonstrate ABI and API stability.

Please let me know if there are any other objectives apart from this.
